### ISSUE TITLE
New logging and cron behavior

### ISSUE THREAD

**Issue by mrworf** [mrworf/plexupdate#137]
To better track this, here's a new issue for this purpose.

To recap, existing logging and cron behavior is not awesome. It started out with the best intentions and good work by @mscreations but it seems to have taken on a life by itself and become unnecessarily kludgy.

So... I've kinda done some work now to improve the situation, please checkout the new branch "reworklog".

The new solution makes away with both quiet and cron options. quiet can be accomplished with >/dev/null and cron is handled by cronwrapper.

Cronwrapper is needed anyway on an ubuntu system since you cannot simply drop plexupdate.sh into any of the cron.* folders. It also adds logging to facility and washing of return codes.

This design also leaves room for doing some nicer handling of a busy server by retrying to apply the update if users are online.

Thoughts? It's kinda jarring so I didn't want to drop it on people. Also need to update the wiki with new details on cronwrapper before merging into master.

@mscreations @demonbane let it rip! 😜

**Comment by demonbane** [comment]
This looks great! It really simplifies the codebase and makes it much easier to understand. I'll try to run some tests on it tomorrow to make sure everything looks good on the cron side.

@mscreations can you double-check the logging functionality and make sure it provides what you were looking for as well? Using syslog for this seems like a great solution.

**Comment by demonbane** [comment]
Ran into a few problems that are addressed in #138. With this, cronwrapper matches the old -c behavior almost perfectly.

**Comment by demonbane** [comment]
What do you think about having cronwrapper read a config file for itself?

I was about to set this up on my machine as a daily job and realized that I can't symlink cronwrapper since I have to edit it, but if I edit it, I won't get any new updates (even though I have autoupdate set).

So I'm thinking something like `~/.plexupdate.cron` that gets read by cronwrapper at runtime for all of its configuration details.

**Comment by mrworf** [comment]
It's not a bad thought, I'm just concerned that we're going to make it overly complicated. The idea behind editing the script was that you copy it and that would be it. 

But in this case, since we're trying to integrate it more with the system, it might make more sense to not have it read it from ```~/.plexupdate.cron``` but ```/etc/plexupdate.cron.conf``` or if it exists, ```/etc/defaults/plexupdate.cron``` My experience is that a lot of non-linux people are installing it and by placing it in /etc we keep it very simple.

Essentially, drop ```cronwrapper``` in ```/etc/cron.weekly``` and edit+drop ```plexupdate.cron``` in ```/etc/defaults``` and you're done

Thoughts?

**Comment by mscreations** [comment]
Finally got a chance to actually test this. It seems to work well enough, but like @demonbane had said, it would be nice if `cronwrapper` had a configuration file for itself.  Not that hard to edit the file, but if `cronwrapper` get's changed due to a bug, you wouldn't get the fix. 

Regarding non-Linux people, what are your thoughts about also providing an installation script? Something that would ask the user for how it should get setup. Then copy the file or symlink it and create a config file in an automated fashion. That way, advanced users could set it up however they want, but more novice users could use the installation script which (hopefully) will set it up correctly without having as much chance of having something go wrong.

**Comment by demonbane** [comment]
I started working on this then get hung up on a decision:

Since we already have `--saveconfig` that essentially functions as an installer for plexupdate, should we extend the built-in functionality of plexupdate to also manage the installation of cronwrapper, or should we pull out saveconfig entirely, and leave all of that to an installer script that can handle both plexupdate and cronwrapper?

**Comment by mrworf** [comment]
@mscreations I like that approach, allows for a fairly easy path for people who don't want to learn linux and keeps the existing implementations fairly clean.

@demonbane No, --saveconfig should not act as an installer, it makes it overly complicated and it is quite unexpect too I think.

Ideally, the installer script can be launched using a one-liner in bash, that checks all dependencies (and for supported platforms, allow you to install necessary dependencies) and then proceeds to git clone it into /opt/plexupdate, ask for config and then setup the cronwrapper. 

Something like:

```
bash -c "$(wget -O - https://raw.githubusercontent.com/mrworf/plexupdate/master/installer.sh)"
```

If all config files are placed in /opt/plexupdate then it's fairly self-contained and we can just add our entry into crontab (of course checking for existance of cronwrapper first). All they really need to be asked is if it's public/plexpass and if plexpass, email and password (which should be tested by the installer before continuing).

Thoughts?

**Comment by demonbane** [comment]
I wasn't suggesting using `--saveconfig` as the installer, just that saveconfig is acting as an "installer" right now. It helps you set up plexupdate the first time and then forget about it, which is exactly what an installer does.

If we're going to use a separate script for installation (which I agree with), I think it makes sense to move the saveconfig functionality into that script rather than leaving it in plexupdate.

That way plexupdate is solely tasked with downloading and installing updates, cronwrapper is solely tasked with running plexupdate on a schedule, and the installer is solely tasked with the initial setup for the various pieces.

**Comment by mscreations** [comment]
@mrworf I was thinking like you were about the one liner for the installer. 

@demonbane I was thinking we should just pull out the `saveconfig` option in favor of using an installer script. 

I envisioned a script that would be interactive and ask things like:

- Where to install plexupdate (default being `/opt/plexupdate` as mrworf suggested.
- Plexpass or Public version
- If Plexpass, then email and username
- Give them a list of other options and allow them to select them to enable or disable options they want
- Then save all those options in the config file, clone into their directory of choice and setup cronwrapper.

**Comment by demonbane** [comment]
@mscreations :+1: That's exactly what I was thinking too. I should have a PR up in a few hours unless you beat me to it. :)

**Comment by mrworf** [comment]
Agreed 😄 sounds like we have a plan of action.

**Comment by mscreations** [comment]
Ha! I won't be that quick. I'm in church at the moment running the sound board... 😃

**Comment by demonbane** [comment]
Just opened a PR for this at #140 and gave both of you write access to my fork so we can work on the PR directly.

I was kind of rushed towards the end of this so I didn't really get a chance to do any cleanup or making sure that everything is a consistent format, so I'll update that tomorrow unless someone wants to pitch in before then.

**Comment by mrworf** [comment]
I took a peek and it looked nice, the only change I made was add validation and make the yesno/noyes functions require a valid response, since I've had some bad experiences personally when scripts assumes default on non-valid inputs 😭 (a slip of the finger and you get some random default, meh, no fun).

Other than that, looks really good :)

Oh yeah, also made it so it skips cron stuff if the /etc/cron.daily is missing (in case of odd arm systems or what have you).

**Comment by demonbane** [comment]
@mrworf That looks great, thanks! I've got 3 VMs set up for testing right now (Ubuntu, Fedora and CentOS) so I'll put everything through its paces in a "real-world" test.

I've also started to think about config file behavior for plexupdate itself. I think the best solution is to have plexupdate look at, in order, `--config`, `/etc/plexupdate.conf`, and `~/.plexupdate`. If `--config` exists we'll stop processing right away, otherwise we'll read each config in order. So if you have both `/etc/plexupdate.conf` and `~/.plexupdate`, the settings in `~/.plexupdate` will take precedence (but they will be "merged" with any settings in `/etc/plexupdate.conf`).

Do you think that's the right behavior or is there a use case I'm missing that might need a different approach? (personally, I'd be happy to **only** use `/etc/plexupdate.conf` but I'm willing to bet that a lot of users still run this as an unprivileged user and wouldn't necessarily want to set it up as root)

**Comment by demonbane** [comment]
I'm testing this right now so pushing a commit that modifies plexupdate to the behavior described above. We can always revert it later if we decide to handle it differently.

**Comment by mrworf** [comment]
Regarding config, I'd expect this (in priority order):

1. --config
2. ~/.plexupdate
3. /etc/plexupdate.conf

So slightly different (swapped 2 & 3), but I think it makes sense, since --config is active override, .plexupdate is your personal setting and /etc/plexupdate.conf is the global setting if all else fails.

**Comment by demonbane** [comment]
So in that case we're only ever using a single configuration file? The way I had it written would first read /etc and then also read .plexupdate if it existed, so it would load both. But that could potentially confuse people I guess.

**Comment by mrworf** [comment]
It will make it quite confusing I think, and also breaks how it works now. If people really wanted that behavior, they could do  ```source /etc/plexupdate.con``` in the top of their personal file 😄

**Comment by mrworf** [comment]
The more I think about it, the more I like dropping .plexupdate since there is soo much code (including eval) which could be scrapped if we did.

**Comment by demonbane** [comment]
So only use --config or /etc/plexupdate.conf and nothing else?

**Comment by mrworf** [comment]
That's what I'm saying... What do you think?

**Comment by mrworf** [comment]
Alright then 😸 let's go with that. We can always provide a legacy plexupdate.conf which restores the original behavior if anyone wants it.

**Comment by demonbane** [comment]
And done: https://github.com/mrworf/plexupdate/pull/140/commits/c6f19afdaa5892aea88ca8824062ce1d4f9816cc

"2 additions & 56 deletions" makes me very happy. :)

**Comment by mrworf** [comment]
Ack, I pushed before I thought it through, I'm just thinking we need to warn existing users about the changed behavior somehow. But the commit I made was bad so I reverted it. 

However, I am concerned that some users will get stuck, but the right place is to document the change in the README.md or I'll end up adding 100+ lines to catch them using the wrong config 👊 😭

**Comment by demonbane** [comment]
We could temporarily put something like:

```bash
if [ -z "${CONFIGFILE}" -a -f ~/.plexupdate -a ! -f /etc/plexupdate.conf ]; then
    echo ".plexupdate has been deprecated, you should move your configuration to /etc/plexupdate.conf"
    echo "Continuing with no config file."
fi
```

and then just take it out after a few weeks?

**Comment by mrworf** [comment]
Yes, good idea, very simple and should catch most which miss the README.md

I'm currently going through the docs and references to .plexupdate so they're gone or explained.

**Comment by mrworf** [comment]
But this also means we've more or less made --save-config redundant, since you can simply rerun the installer?

**Comment by demonbane** [comment]
Yup, that can be removed too. I'm working on a big commit for installer.sh, but if you want to do it go ahead. Otherwise I can do it once I'm done with this.

**Comment by mrworf** [comment]
Go ahead, I'm writing a small novel for the readme ;)

**Comment by mrworf** [comment]
Readme is updated, have a look and correct anything which I got wrong.

**Comment by mrworf** [comment]
Merged into master, closing this issue now

**Comment by Duvrazh** [comment]
Just a thought... you've written in to notify them of deprecation, why not also copy the existing config files to the new location and set them up for seamless upgrades moving forward? Sorry if this seems out of nowhere but I have a vested interest in this script.

**Comment by demonbane** [comment]
@duvrazh this is exactly what the new installer script does. The idea is that you would run installer.sh after the update and it will update everything to make it work as expected with future updates.

Do you think we can communicate this better?

**Comment by mrworf** [comment]
we could add that info to the deprecation message I guess

**Comment by Duvrazh** [comment]
@demonbane that sounds great.  I haven't downloaded the installer and won't be until I'm testing integrating your script with mine for a total pms pia transmission cp sonarr solution.  He'll I'm working on my installer right now.  I like what you're doing.  Keep it up.  Sorry for misunderstanding the comments and taking up your time.

**Comment by hakong** [comment]
Complexity:
We now have a script, a wrapper script for said script, a symlink for the wrapper script, config file(s), and an installer script 😆

Systemd timers/Cron:
Moving from systemd to cron is a step backwards (https://github.com/mrworf/plexupdate/wiki/Running-plexupdate-daily-as-a-systemd-timer/b551bc89894e49effa28738ab5fc74dd575d181c). Systemd units with timers are far superior than cron jobs. See: https://unix.stackexchange.com/questions/278564/cron-vs-systemd-timers and https://wiki.archlinux.org/index.php/Systemd/Timers#Benefits.

Wrapper for automation:
You have a script that's supposed to automate updating of plex media server. Yet it dosen't support automation, but needs a wrapper. Think about that.

Exit logic:
I still don't get the logic of exiting with a non-exit status when plex is up to date with the now removed cron option. You don't see yum-cron, unattended-upgrades, or dnf-automatic exiting with an error when the system is up to date (in addition sending you emails every hour/day/x if run from cron)?

Am I missing, not understanding, or overlooking something or has this gone tits-up?

**Comment by mrworf** [comment]
I disagree, let me explain the reasoning behind all this:

# Complexity
It's no more complex than before. Anyone can just use the plexupdate.sh as they did before. But if you're relatively new to Linux, the installer and cronwrapper makes their life a lot easier. But since it's "fluff", it's not part of plexupdate.sh since it would make it unnecessarily complex. In essence, the cron option and such was broken out to separate parts to keep plexupdate.sh simple.

# Moving away from systemd
We're not moving from systemd to cron, but it's the easiest method for new users. Experienced end-users can use systemd or anything else that they'd like. But by endorsing cron, the confusion is minimized and there is a clear path for people who want fire-and-forget. Cron also works on pretty much all systems and flavors.

Regardless, the systemd entry in the wiki needs help and noone here would be opposed to you adding systemd to the installer and allowing the user a choice or even prefer systemd if available. :)

# Wrapper
Plexupdate.sh was written to workaround the fact that to update I needed to download a deb package, upload it to my server and then run dpkg -i *.deb. It was never intended for cronjobs when I wrote it, and to be frank, I still don't use cron or systemd for it because I cannot have it updating while I'm away from home, getting a borked version of plex and then my wife is unable to use the system. That would not score any points :)

Wrapper allows us to do all kinds of tweaks and adaptations, leading us to...

# exit logic
To be honest, success is defined as:
1. Was able to login and download a NEW version
2. Version wasn't installed already
3. Plex server wasn't in use

If any of these criterias fail, then yes, it's a failure. Now, failure doesn't mean the end of the world, it's in the eye of the beholder. By returning non-zero any automation can use it to determine what happened and take appropriate action. Relying on text output to discern what happened is not good. Also, don't confuse plexupdate.sh with a package manager, But, to your point, it might make sense to kill off error code 5 since there isn't much use to react to it.

And also, cronwrapper can be configured to not send emails :) but I think that might be a hidden gem.

# Summary

You don't have to use installer if you know what you're doing. Please use systemd if you like that and for heavens sake, don't be afraid to submit your systemd work into the repo so people can benefit from it.

In other words, no, I don't feel like it went tits up, and if we can agree that things under "extras" are indeed extras that are optional, then the only real concern you've brought up is the 5 error code which has been a hot potato before on these issue boards before ;) and I'll concede that you're right, it doesn't make much sense to return non-zero if latest version is installed. Expect a patch for this within the hour.

I'll also update  the docs so people using cronwrapper learn how they can disable the email which they get by default.

Sound good to you @hakong ?

**Comment by demonbane** [comment]
Unless I really screwed something up, cronwrapper shouldn't send an email unless an update is actually performed or an error occurs. If everything is up to date there should be no output and, therefore, no email.

**Comment by hakong** [comment]
Sure! Ill make some amendments to the wiki for systemd.
